### DISCUSSION TITLE
Reducing the boilerplate: tapped actions in a list

### DISCUSSION THREAD

**Discussion by tiagoalmeida** [pointfreeco/swift-composable-architecture#379] [General]
First of all let me thank you for the last episode, it shows that you are thinking about **real** problems and usage of TCA.

I have been trying to reduce boilerplate for a while and FormAction/BindingAction is a great place to start!

However, I have a different challenge. I have a settings screen built with UIKit and a TableView that has a couple of options that open other screens. Each option fires an action (i.e : "Team" fires .teamTapped)

## 🧭 Starting the quest 
I have started my quest to reduce boilerplate by wanting to do something like this when I am creating a row that belongs to the SettingsState: 

```.init(title: L10n.Settings.Row.team, action: . teamTapped),```

Then in my reducer I would check: do I have an action? then just return it.

I started that by having a generic in the RowState<Action> and have a property let action: Action. But that always felt wrong to me, and introduces some problems you talked about in the last episode.

## 🤔 Trying to improve what I have with the ideas from last episode

After seeing the episode I was pumped and started to try to improve what I have, but I quickly found out some issues:

- I can't find a good solution to remove this from state (feels wrong to have this in State... for me the State should only contain the data needed to *render* a view)
- Even if I put this in state and try to type erase, I can't give good type ergonomics (such guarantee that is a case from a known action)

Any idea where I could start doing this?

**Comment by mbrandonw** [comment]
Hi @tiagoalmeida, this sounds very interesting, but I think I would need to see more code to understand your situation exactly. Is it possible to whip up a small example with how the state/action/reducer and view code looks?

**Reply by tiagoalmeida** [reply]
Ofc! It is the least I can do.

A lot of simplifications, but something like this: 

[CellTapActionSpike.zip](https://github.com/pointfreeco/swift-composable-architecture/files/5945465/CellTapActionSpike.zip)

**Reply by tiagoalmeida** [reply]
So to make it clear, I am trying to avoid doing things this in the reducer: 

```case rowTapped(1): return openTeams``` .

I am also trying to have a way to declare the action when I build settings (or some sort of model that maps to a tap).

Ideally I could also do something like this in the state initialisation: 

```rows: [.init(title: "Open Team", action: . teamTapped), ...```
